### PR TITLE
Fix the MSVC build

### PR DESCRIPTION
**Pull request by jturcotte**
- Replace functions that are provided by MinGW with a Win32-based
  implementation
- Explicitly export needed symbols from ocsync.dll
- Rename share.h to sharemanager.h since the name clashes with one
  of the Windows headers and get included from there
- Remove the timestamp from the fallback csync stderr logging, it's
  not used since we always provide a log callback

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -166,6 +169,9 @@ find_package(SQLite3 3.8.0 REQUIRED)
 # On some OS, we want to use our own, not the system sqlite
 if (USE_OUR_OWN_SQLITE3)
     include_directories(BEFORE ${SQLITE3_INCLUDE_DIR})
+    if (WIN32)
```
Comment (ogoffart) [inline CMakeLists.txt]: That's both for MSVC and mingw,  right?

File: cmake/modules/Warnings.cmake [hunk 2]
```diff
@@ -2,21 +2,23 @@
 # Redistribution and use is allowed according to the terms of the BSD license.
 # For details see the accompanying COPYING* file.
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -Wno-long-long")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x")
+if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -Wno-long-long")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x")
 
-if(CMAKE_COMPILER_IS_GNUCXX)
```
Comment (ogoffart) [inline cmake/modules/Warnings.cmake]: oh, was that not more correct?

File: csync/src/csync_exclude.c [hunk 3]
```diff
@@ -102,7 +107,7 @@ int csync_exclude_load(const char *fname, c_strlist_t **list) {
   int fd = -1;
   int i = 0;
   int rc = -1;
-  int64_t size;
+  off_t size;
```
Comment (ogoffart) [inline csync/src/csync_exclude.c]: We have had problems with off_t before  (see commit 4f5bfbdb784fb4c589ff04c439fb82749b3b54e5 )

File: csync/src/csync_log.c [hunk 4]
```diff
@@ -37,46 +31,11 @@ CSYNC_THREAD int csync_log_level;
...
-    int rc;
-
-    rc = current_timestring(1, date, sizeof(date));
-    if (rc == 0) {
-        fprintf(stderr, "[%s, %d] %s:", date+5, verbosity, function);
-    } else {
-        fprintf(stderr, "[%d] %s", verbosity, function);
-    }
-
+    fprintf(stderr, "[%d] %s", verbosity, function);
```
Comment (ogoffart) [inline csync/src/csync_log.c]: Yes we don"t really need the date...
Then the two fprintf can be merged.

File: csync/src/std/asprintf.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (ogoffart) [inline csync/src/std/asprintf.c]: Should the file go in 3rdparty?
Should there be a URL from the readme on where this came from, and which revision. ?

File: CMakeLists.txt (hunk 1 above)
Comment (jturcotte) [inline CMakeLists.txt]: Yes, same for _WIN32 in the code, it applies to both MSVC and MinGW

File: cmake/modules/Warnings.cmake (hunk 2 above)
Comment (jturcotte) [inline cmake/modules/Warnings.cmake]: They just recommend to use the compiler ID if we want to target both Clang and GCC, my last commit to that file was getting stuff out of that block for Clang, but now I need to keep MSVC out of it.

File: csync/src/csync_exclude.c (hunk 3 above)
Comment (jturcotte) [inline csync/src/csync_exclude.c]: OK that one was only a warning so I'll remove it.

File: csync/src/std/asprintf.c (hunk 5 above)
Comment (jturcotte) [inline csync/src/std/asprintf.c]: It's unlikely we'll update that code ever again so I think it's not worth it to start referencing files across the filesystem to 3rdparty in the cmake files.

Where should I put the URL, in the file itself?

File: csync/src/csync_exclude.c (hunk 3 above)
Comment (jturcotte) [inline csync/src/csync_exclude.c]: done

File: csync/src/std/asprintf.c (hunk 5 above)
Comment (jturcotte) [inline csync/src/std/asprintf.c]: I've put the URL in the file headers.

**Comment by jturcotte** [comment]
Please look at the commit message for the reason why renaming share.h was necessary.

**Comment by ogoffart** [comment]
:+1:

**Pull request merged**